### PR TITLE
pythonPackages.affine: init at 2.2.1

### PR DESCRIPTION
**Pull request by mredaelli**
cartopy: init at 0.17.0
pyepsg: init at 0.3.2
rasterio: init at 0.34.0
snug: init at 1.3.4
snuggs: init at 1.4.2

A few libraries related to geospatial data analysis

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by FRidh** [comment]
please rebase instead of merging

**Comment by mredaelli** [comment]
Ack

**Comment by mredaelli** [comment]
@c0bw3b Sorry, do you know why earlier it was failing while now, after your message, it worked?
Did I do something wrong?

**Comment by c0bw3b** [comment]
@mredaelli no you did nothing wrong, it was a temporary breakage in `master` branch

File: pkgs/development/python-modules/cartopy/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cartopy/default.nix]:
```suggestion
```

File: pkgs/development/python-modules/cartopy/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cartopy/default.nix]:
```suggestion
  pname = "cartopy";
```

See https://nixos.org/nixpkgs/manual/#sec-package-naming

File: pkgs/development/python-modules/rasterio/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/rasterio/default.nix]:
```suggestion
  numpy, affine, cligj, click-plugins, snuggs, gdal
```

File: pkgs/top-level/python-packages.nix
Comment (worldofpeace) [inline pkgs/top-level/python-packages.nix]: Libraries in pkgs/top-level/python-packages.nix are sorted quasi-alphabetically to avoid merge conflicts.

File: pkgs/development/python-modules/cartopy/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cartopy/default.nix]: This description can be shortened for brevity.

Something like

"Process geospatial data to create maps and perform analyses"

File: pkgs/development/python-modules/cartopy/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cartopy/default.nix]:
```suggestion
    # optional
```

File: pkgs/top-level/python-packages.nix
Comment (worldofpeace) [reply pkgs/top-level/python-packages.nix]: Can you try to sort these through the entirety of this file as alphabetical as you can?

Currently you've just alphabetically sorted the packages you've added at the very end of the file, which is not what we want.

Comment (mredaelli) [reply pkgs/top-level/python-packages.nix]: Sure, I'll try, but again, they sorting restarts several times throughout the file.
Let's see :)

Comment (mredaelli) [reply pkgs/top-level/python-packages.nix]: OK, give it another try

File: pkgs/top-level/python-packages.nix
Comment (worldofpeace) [inline pkgs/top-level/python-packages.nix]: There's whitespace here from before.

**Comment by worldofpeace** [comment]
Also looks like accidentally squashed your commits.

**Comment by worldofpeace** [comment]
Going to check if they build properly now since my feedback isn't critical to their function.

@GrahamcOfBorg pythonPackages.cartopy python3Packages.cartopy pythonPackages.pyepsg python3Packages.pyepsg pythonPackages.rasterio python3Packages.rasterio pythonPackages.snug python3Packages.snug pythonPackages.snuggs python3Packages.snuggs

**Comment by mredaelli** [comment]
Cartopy works only with Python 3.6. Not sure if that's a problem, or how to manage it

**Comment by worldofpeace** [comment]
> Cartopy works only with Python 3.6. Not sure if that's a problem, or how to manage it

`buildPythonPackages` takes a parameter `disabled` when true, it won't build for that python interpreter version.

However from https://github.com/SciTools/cartopy/issues/1186 I can see that it does work for other interpreter versions, though they are looking to drop support for python2 very soon.

Shouldn't be a problem yet.

**Comment by mredaelli** [comment]
It's the (optional) dependency on `pyproj` through `owslib`, which doesn't work: https://github.com/jswhit/pyproj/issues/136

Should I add a withOwslib parameter for the derivation?

**Comment by worldofpeace** [comment]
> Should I add a withOwslib parameter for the derivation?

This should be fine

```diff
diff --git a/pkgs/development/python-modules/cartopy/default.nix b/pkgs/development/python-modules/cartopy/default.nix
index 4c6e05149fc..efe26b217fa 100644
--- a/pkgs/development/python-modules/cartopy/default.nix
+++ b/pkgs/development/python-modules/cartopy/default.nix
@@ -1,6 +1,6 @@
 { buildPythonPackage, lib, fetchPypi,
   pytest, filelock, mock, pep8,
-  cython,
+  cython, isPy37,
   six, pyshp, shapely, geos, proj, numpy,
   gdal, pillow, matplotlib, pyepsg, pykdtree, scipy, owslib, fiona
 }:
@@ -24,8 +24,8 @@ buildPythonPackage rec {
     six pyshp shapely geos proj numpy
 
     # optional
-    gdal pillow matplotlib pyepsg pykdtree scipy owslib fiona
-  ];
+    gdal pillow matplotlib pyepsg pykdtree scipy fiona
+  ] ++ lib.optionals (!isPy37) [ owslib ]; # owslib is broken with python3.7
 
   meta = with lib; {
     description = ''Process geospatial data to create maps and perform analyses.'';

```

**Comment by worldofpeace** [comment]
Also noticed this in the build output for `cartopy`

`Ran 0 tests in 0.000s`

I don't think the tests are being ran.

File: pkgs/development/python-modules/rasterio/default.nix
Comment (alyssais) [inline pkgs/development/python-modules/rasterio/default.nix]: Commas should be at the start, not end, of lines. (See the [manual](https://nixos.org/nixpkgs/manual/#sec-syntax).)

**Comment by mredaelli** [comment]
@worldofpeace true, I added them, and disabled the ones that require network access

File: pkgs/development/python-modules/cartopy/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cartopy/default.nix]:
```suggestion
    description = "Process geospatial data to create maps and perform analyses"';
```
Double quotes are fine here, and we'd don't use periods at the end of the description.

Comment (mredaelli) [reply pkgs/development/python-modules/cartopy/default.nix]: Done commas, dot and quotes

File: pkgs/development/python-modules/cartopy/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cartopy/default.nix]: I'm seeing a warning `"Could not set locale to English/United States. "` during the tests, so I think we should set the locale like `snug`.

**Comment by worldofpeace** [comment]
@GrahamcOfBorg pythonPackages.cartopy python3Packages.cartopy pythonPackages.pyepsg python3Packages.pyepsg pythonPackages.rasterio python3Packages.rasterio pythonPackages.snug python3Packages.snug pythonPackages.snuggs python3Packages.snuggs pythonPackages.affine python3Packages.affine

File: pkgs/development/python-modules/snug/default.nix
Comment (mredaelli) [reply pkgs/development/python-modules/snug/default.nix]: They were not in the pypi package. I switched to github, and then had to add a new package that is needed for the tests.

Comment (mredaelli) [reply pkgs/development/python-modules/snug/default.nix]: Oh, I used isPy35 and isPy34, too, even though I don't think those versions are even available. Hope that's OK

File: pkgs/development/python-modules/snug/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/snug/default.nix]:
```nix
lib.optionals (pythonOlder "3.4") [ singledispatch ] ++
lib.optionals (pythonOlder "3.5") [ typing ];
```

Using [`pythonOlder`](https://github.com/NixOS/nixpkgs/blob/6dbf15d627ecd1a18bf965000c86b63b70629ac6/pkgs/top-level/python-packages.nix#L22) feels nicer here.

File: pkgs/development/python-modules/snug/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/snug/default.nix]:
```suggestion
  checkPhase = "pytest";
```

File: pkgs/development/python-modules/snug/default.nix
Comment (mredaelli) [reply pkgs/development/python-modules/snug/default.nix]: Absolutely. I didn't know that, thanks.

File: pkgs/development/python-modules/snug/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/snug/default.nix]:
```suggestion
, pythonOlder
```

**Comment by worldofpeace** [comment]
@mredaelli Every addition of a package needs to be a separate commit.

File: pkgs/development/python-modules/gentools/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/gentools/default.nix]: Please do the same as what has been done in `snug` with `pythonOlder`.

File: pkgs/development/python-modules/gentools/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/gentools/default.nix]:
```suggestion
```
Be sure to remove these after the below is applied and add `pythonOlder`.

Comment (mredaelli) [reply pkgs/development/python-modules/gentools/default.nix]: Ops, shouldn't commit early in the morning ^_^

File: pkgs/development/python-modules/snuggs/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/snuggs/default.nix]: This test wasn't being ran https://github.com/mapbox/snuggs/blob/master/test_snuggs.py

File: pkgs/development/python-modules/cartopy/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cartopy/default.nix]: The old description sneaked in somehow.

File: pkgs/development/python-modules/snuggs/default.nix
Comment (mredaelli) [reply pkgs/development/python-modules/snuggs/default.nix]: Fetched from github and added

File: pkgs/development/python-modules/cartopy/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/cartopy/default.nix]: WIll probably be fixed by https://github.com/NixOS/nixpkgs/pull/50982.

File: pkgs/development/python-modules/rasterio/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/rasterio/default.nix]: Why not `fetchPypi`? Please add a comment.

File: pkgs/development/python-modules/snuggs/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/snuggs/default.nix]: Add a comment on why you're using `fetchFromGitHub`.

**Review by dotlambda — CHANGES_REQUESTED** [review]
I'd prefer beginning the commit messages with `pythinPackages.<attrname>` but that's up to you.

File: pkgs/development/python-modules/snuggs/default.nix
Comment (mredaelli) [reply pkgs/development/python-modules/snuggs/default.nix]: Fixed

File: pkgs/development/python-modules/cartopy/default.nix
Comment (dotlambda) [reply pkgs/development/python-modules/cartopy/default.nix]: You should be able to rebase on master and get rid of this.

Comment (mredaelli) [reply pkgs/development/python-modules/cartopy/default.nix]: Yep, locally it worked both for python 3.6 and 3.7.

Comment (mredaelli) [reply pkgs/development/python-modules/cartopy/default.nix]: Ready for review :)

File: pkgs/development/python-modules/rasterio/default.nix
Comment (alyssais) [reply pkgs/development/python-modules/rasterio/default.nix]: @worldofpeace you marked this as resolved, but unless GitHub is being broken, it doesn't seem to have been fixed?

**Comment by worldofpeace** [comment]
@dotlambda 
> I'd prefer beginning the commit messages with pythinPackages.<attrname> but that's up to you.

I'm seeing a lot of different commit msg for python so I don't even know what to suggest to people :smile:

File: pkgs/development/python-modules/cartopy/default.nix
Comment (worldofpeace) [reply pkgs/development/python-modules/cartopy/default.nix]: Actually opened that to fix this :)

File: pkgs/development/python-modules/snug/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/snug/default.nix]: Forgot that a comment here for why we need to set locale would be helpful.

If I recall correctly, it was to silence a warning while the tests were being ran.

File: pkgs/development/python-modules/cartopy/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cartopy/default.nix]: Also forgot to suggest here, a comment (if you can recall) why this was needed is helpful to others.

File: pkgs/development/python-modules/cartopy/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cartopy/default.nix]: Also forgot to suggest here, a comment (if you can recall) why this was needed is helpful to others.

**Comment by alyssais** [comment]
> > I'd prefer beginning the commit messages with pythinPackages. but that's up to you.
> 
> I'm seeing a lot of different commit msg for python so I don't even know what to suggest to people 😄

`pythonPackages.whatever: ` commit messages mean that ofborg will automatically build them (if you have that access), so recommending that one to everybody is probably most sensible for consistency…

**Review by worldofpeace — DISMISSED** [review]
I've successfully built all of the packages locally(both python attributes).

Everything looks great to me, however we should wait if @dotlambda has any other feedback.

File: pkgs/development/python-modules/cartopy/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/cartopy/default.nix]:
```suggestion
  pname = "Cartopy";
```

File: pkgs/development/python-modules/cartopy/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/cartopy/default.nix]:
```suggestion
```

File: pkgs/development/python-modules/affine/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/affine/default.nix]: You might want to add
```
checkPhase = ''
  py.test
'';
```
because that runs 62 instead of only 46 tests and it's what's upstream uses.

File: pkgs/development/python-modules/snug/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/snug/default.nix]: You're sure this is necessary? It builds fine for me without.

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/snug/default.nix
Comment (mredaelli) [reply pkgs/development/python-modules/snug/default.nix]: It certainly was when I was fetching from Pypi. I will take a look

File: pkgs/development/python-modules/affine/default.nix
Comment (mredaelli) [reply pkgs/development/python-modules/affine/default.nix]: Sure, will do that

File: pkgs/development/python-modules/cartopy/default.nix
Comment (mredaelli) [reply pkgs/development/python-modules/cartopy/default.nix]: Ehm, this would basically undo another change that was requested earlier by @worldofpeace .
It starts to feel like this review is not converging :D

Comment (dotlambda) [reply pkgs/development/python-modules/cartopy/default.nix]: You're referring to https://github.com/NixOS/nixpkgs/pull/50951#discussion_r236031890. I must admit I probably never read that section of the manual. Personally, I don't think package names are of any importance compared to attribute names and I'd still prefer a single `pname`. What does @FRidh think?

File: pkgs/development/python-modules/snug/default.nix
Comment (mredaelli) [reply pkgs/development/python-modules/snug/default.nix]: Without it, I still get
```
  File "setup.py", line 21, in <module>
    long_description=read('README.rst') + '\n\n' + read('HISTORY.rst'),
  File "setup.py", line 9, in read
    return rfile.read()
  File "/nix/store/yk7hci7pg3n2kyjh8il317wvd5nam3aw-python3-3.6.7/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xf0 in position 5: ordinal not in range(128)
```

Comment (dotlambda) [reply pkgs/development/python-modules/snug/default.nix]: That's strange: Without the locale setting, `python37.pkgs.snug` builds fine while `python36.pkgs.snug` doesn't. Let's just keep it in.

File: pkgs/development/python-modules/snug/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/snug/default.nix]:
```suggestion
  # Prevent unicode decoding error in setup.py
```

File: pkgs/development/python-modules/cartopy/default.nix
Comment (dotlambda) [reply pkgs/development/python-modules/cartopy/default.nix]: Judging by https://github.com/costrouc/nixpkgs-dev-tools/pull/3/commits/1953ac72be9cd4fdc8ee31be16c7999812c91a15, it seems like we should indeed normalize `pname`s.

**Review by worldofpeace — APPROVED** [review]
@mredaelli Again, I'm able to build these all successfully :+1: 

I think this is good to merge. 
I'll go ahead and do so once we have @dotlambda approval.

**Comment by worldofpeace** [comment]
@mredaelli Thanks for contributing and bearing with us when our reviews weren't converging :smile:

**Comment by mredaelli** [comment]
@worldofpeace Not at all! It was very instructive, thank you all

**Pull request merged**
